### PR TITLE
AWS Query protocol to Json Protcol migration compatibility, support for awsQueryError metadata trait

### PR DESCRIPTION
**Pull request by ziyanli-amazon**
- Parse error code if it is present in HTTP response header

EDIT: removed the link to SEP - _debora-ito_.

## Motivation and Context

As we decided to parse AwsQuery Compatible Error in the Http response header, the required changes need to take in place on both SDK client and service sides. This is the SDK client needed change. 
This will be a pre-requisite for migrating services from AWSQuery wire protocol to AWSJson.

## Modifications

Error code can be fetched from a http header (for example `x-amzn-ErrorType`, `:error-code`, or `:exception-type`) or content (for example `__type#fooException`). This PR takes another scenario into consideration, of which when `x-amzn-query-error` header with valid value is found from the Http response. The parsed error code remains unchanged if this header is missing, contains `null` or invalid value. 

## Testing

Stemmed from the SEP and i did a brief extension on test scenarios to 
- TC1: Validate SDK can parse Code field from the Http header `x-amzn-query-error`
- TC2: Validate SDK can correctly parse Code field when `x-amzn-query-error` is null
- TC3: Validate SDK can correctly parse Code field when `x-amzn-query-error` is empty
- TC4: Validate SDK can correctly parse Code field when `x-amzn-query-error` is invalid

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: .changes/next-release/feature-core-cced62e.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "type": "feature",
+    "category": "core",
+    "contributor": "ziyanli-amazon",
+    "description": "As we decided to parse AwsQuery Compatible Error in the Http response header, the required changes need to take in place on both SDK client and service sides. This is the SDK client needed change.\nThis will be a pre-requisite for migrating services from AWSQuery wire protocol to AWSJson."
```
Comment (joviegas) [inline .changes/next-release/feature-core-cced62e.json:5]: Support  for clients with AWS Query protocol migrating towards AwsJson in order to  supports customizing error codes.

**Comment by joviegas** [comment]
1. We will need codegen changes to identify services with @awsQueryCompatible trait.
2. Treat `x-amzn-query-error`  header in specail way onlt for clients that are marked with @awsQueryCompatible .
3. Protocol test needs to be upated

File: codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/Metadata.java [hunk 2]
```diff
@@ -499,6 +504,17 @@ public Metadata withJsonVersion(String jsonVersion) {
         return this;
     }
 
+    public Map<String, String> getAwsQueryCompatible() { return awsQueryCompatible; }
+
+    public void setAwsQueryCompatible(Map<String, String> awsQueryCompatible) {
+        this.awsQueryCompatible = awsQueryCompatible;
+    }
+
+    public Metadata withAwsQueryCompatible(Map<String, String> awsQueryCompatible) {
```
Comment (joviegas) [inline codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/Metadata.java:513]: Was looking for some confirmation that awsQueryCompatible is of key-value type String-String, the specification doesnot define the type for Map

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/specs/JsonProtocolSpec.java [hunk 3]
```diff
@@ -94,7 +94,8 @@ public MethodSpec initProtocolFactory(IntermediateModel model) {
                       .addCode(".defaultServiceExceptionSupplier($T::builder)\n", baseException)
                       .addCode(".protocol($T.$L)\n", AwsJsonProtocol.class, protocolEnumName(metadata.getProtocol()))
                       .addCode(".protocolVersion($S)\n", metadata.getJsonVersion())
-                      .addCode("$L", customErrorCodeFieldName());
+                      .addCode("$L", customErrorCodeFieldName())
+                      .addCode("$L", hasAwsQueryCompatible());
```
Comment (joviegas) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/specs/JsonProtocolSpec.java:98]: can we conditionally add this only if hasAwsQueryCompatible is true similar to V1 ?

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/json/service-2.json [hunk 4]
```diff
@@ -11,7 +11,8 @@
     "signingName": "json-service",
     "signatureVersion": "v4",
     "uid": "json-service-2010-05-08",
-    "xmlNamespace": "https://json-service.amazonaws.com/doc/2010-05-08/"
+    "xmlNamespace": "https://json-service.amazonaws.com/doc/2010-05-08/",
+    "awsQueryCompatible": {}
```
Comment (joviegas) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/json/service-2.json:15]: Can we add a separate file here to test this since this is specific to only one service now and this will fail to test when  "awsQueryCompatible": {} is not existent in c2j model

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java [hunk 5]
```diff
@@ -87,6 +91,20 @@ private AwsServiceException unmarshall(SdkHttpFullResponse response, ExecutionAt
         return exception.build();
     }
 
+    private String getEffectiveErrorCode(SdkHttpFullResponse response, String errorCode) {
```
Comment (joviegas) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java:94]: Can we do this.hasAwsQueryCompatible at the beginning , can we break this function into smaller functions just as commented for V1 ?

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java [hunk 6]
```diff
@@ -218,5 +237,17 @@ public Builder errorCodeParser(ErrorCodeParser errorCodeParser) {
         public AwsJsonProtocolErrorUnmarshaller build() {
             return new AwsJsonProtocolErrorUnmarshaller(this);
         }
+
+        /**
+         * Provides a check on whether AwsQueryCompatible trait is found in Metadata.
+         * If true, custom error codes can be provided
```
Comment (joviegas) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java:243]: Can we add more clarification like `Error code is derived from custom header` , or something like that specific to this.

File: core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java [hunk 7]
```diff
@@ -55,6 +58,27 @@ public void unmarshal_faultValue_useCorrectly(TestCase tc) {
         AwsServiceException exception = unmarshaller.handle(responseBuilder.build(), new ExecutionAttributes());
 
         assertThat(exception.statusCode()).isEqualTo(tc.expectedStatusCode);
+        assertThat(exception.awsErrorDetails().errorCode()).isEqualTo("ServiceException");
+    }
+
+    @ParameterizedTest
+    @MethodSource("x_amzn_query_error_testCases")
```
Comment (joviegas) [inline core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java:65]: 👍

File: core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java [hunk 8]
```diff
@@ -66,7 +90,17 @@ public static List<TestCase> unmarshal_faultValue_testCases() {
         );
     }
 
-    private static AwsJsonProtocolErrorUnmarshaller makeUnmarshaller(List<ExceptionMetadata> exceptionMetadata) {
+    public static List<QueryErrorTestCase> x_amzn_query_error_testCases() {
+        return Arrays.asList(
+            new QueryErrorTestCase("actualErrorCode;Sender", "actualErrorCode"),
+            new QueryErrorTestCase("actualError CodeSender", "ServiceException"),
+            new QueryErrorTestCase("actualError", "ServiceException"),
```
Comment (joviegas) [inline core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java:97]: Just a query , why delimeter ';' so much important here , why cant we take the error message whatever we find when delimeter doesnot exist like `actualError` in this case.
Anyways I donot see the text following the delimter is used by SDK in this case ;Sender the Sender is used not used anywhere.

**Comment by joviegas** [comment]
1. Please add test cases in codegen-generated-classes-test with service2.json  containing `awsQueryCompatible` something similar to src/main/resources/codegen-resources/bearerauth/service-2.json
2. Please add a test case in PoetClientFunctionalTests in codegen/src/test/java/software/amazon/awssdk/codegen/poet/client/PoetClientFunctionalTests.java refer to

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/specs/JsonProtocolSpec.java (hunk 3 above)
Comment (joviegas) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/specs/JsonProtocolSpec.java:98]: Please run unit test case codegen module , this change would have cause failures in existing test cases

File: codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/Metadata.java (hunk 2 above)
Comment (ziyanli-amazon) [reply codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/Metadata.java:513]: since it is an empty map anyway, string string map should suffice.

File: core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java (hunk 8 above)
Comment (ziyanli-amazon) [reply core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java:97]: Yes, I'm just following the rule made by the SEP. the semi colon is a hard requirement. So if the header only contains the errorCode (without the delimiter or error type), sdk should fall back parsing the errorCode from its original source (either from content or something else). The `;Sender` is indeed not used in Java sdk but in some other sdk, it does exposed to the customer.

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java (hunk 6 above)
Comment (ziyanli-amazon) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java:243]: Sure. will add that.

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java [hunk 9]
```diff
@@ -87,6 +92,35 @@ private AwsServiceException unmarshall(SdkHttpFullResponse response, ExecutionAt
...
+            String compatibleErrorCode = queryCompatibleErrorCodeFromResponse(response);
+            if (!StringUtils.isEmpty(compatibleErrorCode)) {
+                return compatibleErrorCode;
+            }
+        }
+        return errorCode;
+    }
+
+    private String queryCompatibleErrorCodeFromResponse(SdkHttpFullResponse response) {
+        List<String> headerValues = response.headers().get(X_AMZN_QUERY_ERROR);
```
Comment (joviegas) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java:106]: Please use
```
        Optional<String> headerValue = response.firstMatchingHeader(X_AMZN_QUERY_ERROR);
        if (headerValue.isPresent()) {
```

File: codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/Metadata.java (hunk 2 above)
Comment (joviegas) [reply codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/Metadata.java:519]: It would be helpful in feature if we mention this in spec document.

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/json-query-error/customization.config [hunk 10]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (joviegas) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/json-query-error/customization.config:1]: Can we rename this folder to
query-to-json-errorcode

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java (hunk 9 above)
Comment (ziyanli-amazon) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java]: changed

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/json-query-error/customization.config (hunk 10 above)
Comment (ziyanli-amazon) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/json-query-error/customization.config]: done

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/awsquerycompatible/AwsQueryCompatibleErrorTest.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.querycompatiblejson.QueryCompatibleJsonAsyncClient;
+import software.amazon.awssdk.services.querycompatiblejson.QueryCompatibleJsonClient;
+import software.amazon.awssdk.services.querycompatiblejson.model.AllTypeResponse;
+import software.amazon.awssdk.services.querycompatiblejson.model.QueryCompatibleJsonResponse;
+import software.amazon.awssdk.utils.builder.SdkBuilder;
+
+public class AwsQueryCompatibleErrorTest {
```
Comment (joviegas) [inline test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/awsquerycompatible/AwsQueryCompatibleErrorTest.java:38]: Can we add test cases with different queryHeaderValue like one with some  spaces before and after the value , you had mentioned we no longer need trimming.
Can we add some cases where queryHeaderValue can have unexpected values

File: test/codegen-generated-classes-test/src/main/resources/codegen-resources/awsquerycompatible/customization.config [hunk 12]
```diff
@@ -0,0 +1,6 @@
+{
+    "customResponseMetadata": {
+        "X_AMZN_QUERY_ERROR": "x-amzn-query-error"
```
[the remark this answers is not in the document]
Comment (ziyanli-amazon) [reply test/codegen-generated-classes-test/src/main/resources/codegen-resources/awsquerycompatible/customization.config:3]: This is to mock the header value getting parsed. This doesn't need to be included in any service customization.config. I believe coral is responsible for populating that value.

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java [hunk 13]
```diff
@@ -41,12 +41,14 @@
 @SdkInternalApi
 public final class AwsJsonProtocolErrorUnmarshaller implements HttpResponseHandler<AwsServiceException> {
 
+    private static final String QUERY_ERROR_DELIMITER = ";";
```
[the remark this answers is not in the document]
Comment (ziyanli-amazon) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/unmarshall/AwsJsonProtocolErrorUnmarshaller.java:44]: Will do

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/query-to-json-errorcode/service-2.json [hunk 14]
```diff
@@ -0,0 +1,512 @@
+{
+  "version": "2.0",
+  "metadata": {
+    "apiVersion": "2010-05-08",
+    "endpointPrefix": "json-service-endpoint",
+    "globalEndpoint": "json-service.amazonaws.com",
+    "protocol": "json",
+    "serviceAbbreviation": "Json Service",
```
[the remark this answers is not in the document]
Comment (ziyanli-amazon) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/query-to-json-errorcode/service-2.json:8]: Ok.

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/query-to-json-errorcode/service-2.json [hunk 15]
```diff
@@ -0,0 +1,512 @@
...
+    "serviceFullName": "Some Service That Uses AWS Json Protocol",
+    "serviceId": "Json Service",
+    "signingName": "json-service",
+    "signatureVersion": "v4",
+    "uid": "json-service-2010-05-08",
+    "xmlNamespace": "https://json-service.amazonaws.com/doc/2010-05-08/",
+    "awsQueryCompatible": {}
+  },
+  "operations": {
```
[the remark this answers is not in the document]
Comment (ziyanli-amazon) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/query-to-json-errorcode/service-2.json:17]: No. I simply copied and pasted. Will remove unwanted operations and only keep one

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/query-to-json-errorcode/customization.config [hunk 16]
```diff
@@ -0,0 +1,24 @@
+{
```
[the remark this answers is not in the document]
Comment (ziyanli-amazon) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/query-to-json-errorcode/customization.config:1]: Done

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/awsquerycompatible/AwsQueryCompatibleErrorTest.java (hunk 11 above)
Comment (ziyanli-amazon) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/awsquerycompatible/AwsQueryCompatibleErrorTest.java:38]: So the value of these tests I added here is to mock end-to-end codegen logic for the new `AwsQueryCompatible` trait. As the codegen doesn't control or verify the validity of the header value, I don't see any value or know how to implement the requested test scenarios. The requested test scenarios were already addressed in the `aws-json-protocol` parameterized tests. Let me know if you have a different opinion.

File: core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java [hunk 17]
```diff
@@ -66,7 +90,17 @@ public static List<TestCase> unmarshal_faultValue_testCases() {
...
+            new QueryErrorTestCase(";Sender", "ServiceException")
```
Comment (zoewangg) [inline core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java:98]: Can we add a test case for empty value as well as null value?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/http/HttpResponseHandler.java [hunk 18]
```diff
@@ -40,6 +40,7 @@
                                                                                   X_AMZN_REQUEST_ID_HEADER_ALTERNATE)
                                                                               .collect(Collectors.toSet()));
     String X_AMZ_ID_2_HEADER = "x-amz-id-2";
+    String X_AMZN_QUERY_ERROR = "x-amzn-query-error";
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/http/HttpResponseHandler.java:43]: Do we need it here?

File: .changes/next-release/feature-core-cced62e.json [hunk 19]
```diff
@@ -0,0 +1,6 @@
+{
+    "type": "feature",
+    "category": "core",
```
Comment (zoewangg) [inline .changes/next-release/feature-core-cced62e.json:3]: nit, core changes should be `AWS SDK for Java v2`

File: codegen/src/test/java/software/amazon/awssdk/codegen/poet/client/PoetClientFunctionalTests.java [hunk 20]
```diff
@@ -75,6 +75,12 @@ public void asyncClientClassAwsJson() throws Exception {
         assertThat(asyncClientClass, generatesTo("test-aws-json-async-client-class.java"));
     }
 
+    @Test
+    public void asyncClientClassAwsQueryCompatibleJson() throws Exception {
```
Comment (zoewangg) [inline codegen/src/test/java/software/amazon/awssdk/codegen/poet/client/PoetClientFunctionalTests.java:79]: Can we add a test for sync client?

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/test-aws-query-compatible-json-async-client-class.java [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+        clientHandler.close();
+    }
+
+    private <T extends BaseAwsJsonProtocolFactory.Builder<T>> T init(T builder) {
+        return builder
+            .clientConfiguration(clientConfiguration)
+            .defaultServiceExceptionSupplier(QueryToJsonCompatibleException::builder)
+            .protocol(AwsJsonProtocol.AWS_JSON)
+            .protocolVersion("1.1")
+            .hasAwsQueryCompatible(true)
```
Comment (zoewangg) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/test-aws-query-compatible-json-async-client-class.java:143]: Is `.hasAwsQueryCompatible(true)` the only code being added? Could you show me a diff between client w/o hasAwsQueryCompatible and client w/ hasAwsQueryCompatible?

Comment (ziyanli-amazon) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/test-aws-query-compatible-json-async-client-class.java:143]: Yes, `.hasAwsQueryCompatible(true)` is the only code being added. This `test-aws-query-compatible-json-async-client-class.java` file was auto generated based on service-2.json file added here: https://github.com/ziyanli-amazon/aws-sdk-java-v2/blob/master/test/codegen-generated-classes-test/src/main/resources/codegen-resources/awsquerycompatible/service-2.json and this service-2.json is a copy and paste of https://github.com/ziyanli-amazon/aws-sdk-java-v2/blob/master/test/codegen-generated-classes-test/src/main/resources/codegen-resources/bearerauth/service-2.json with `awsQueryCompatible` field added.

File: .changes/next-release/feature-core-cced62e.json (hunk 19 above)
Comment (ziyanli-amazon) [reply .changes/next-release/feature-core-cced62e.json:3]: I see. Thanks. I was following the instructions in the README. Will change it.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/http/HttpResponseHandler.java (hunk 18 above)
Comment (ziyanli-amazon) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/http/HttpResponseHandler.java:43]: Yes, I'm trying to follow the coding convention of the package here. i saw the other headers which are used by the same ErrorHandler class reside here. Thought it is a good idea. Let me know if you think it is better put it elsewhere.

Comment (zoewangg) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/http/HttpResponseHandler.java:43]: Can we move it to the class that uses it? It seems to be very specific to query protocol and this class is a generic handler class

Comment (ziyanli-amazon) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/http/HttpResponseHandler.java:43]: ok. will do

File: core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java (hunk 17 above)
Comment (ziyanli-amazon) [reply core/protocols/aws-json-protocol/src/test/java/software/amazon/awssdk/protocols/json/FaultStatusCodeMappingTest.java:98]: sure. will do.

File: codegen/src/test/java/software/amazon/awssdk/codegen/poet/client/PoetClientFunctionalTests.java (hunk 20 above)
Comment (ziyanli-amazon) [reply codegen/src/test/java/software/amazon/awssdk/codegen/poet/client/PoetClientFunctionalTests.java:79]: sure. will add it.

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/awsquerycompatible/AwsQueryCompatibleErrorTest.java [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+        assertThat(allTypesResponse.responseMetadata().xAmznQueryError()).isEqualTo(queryHeaderValue);
+    }
+
+    private void verifyUnknownResponseMetadata(QueryCompatibleJsonResponse allTypesResponse) {
+        assertThat(allTypesResponse.responseMetadata()).isNotNull();
+        assertThat(allTypesResponse.responseMetadata().xAmznQueryError()).isEqualTo("UNKNOWN");
+    }
+
+    private void stubResponseWithHeaders(String headerValue) {
+        stubFor(post(anyUrl())
```
[the remark this answers is not in the document]
Comment (ziyanli-amazon) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/awsquerycompatible/AwsQueryCompatibleErrorTest.java:106]: Yes, modified in the latest commit

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/awsquerycompatible/AwsQueryCompatibleErrorTest.java [hunk 23]
```diff
@@ -62,59 +63,58 @@ public void setupClient() {
...
 
     @Test
-    public void asyncSubWithHeader_shouldContainResponseMetadata() {
-        stubResponseWithHeaders(queryHeaderValue);
-        AllTypeResponse allTypeAsyncResponse = asyncClient.allType(SdkBuilder::build).join();
-        verifyResponseMetadata(allTypeAsyncResponse, queryHeaderValue);
+    public void verifyAsyncClientException_shouldRetrievedFromHeader() {
+        stubResponseWithQueryHeaderAndBody(QUERY_HEADER_VALUE);
+        try {
+            asyncClient.allType(SdkBuilder::build);
```
Comment (zoewangg) [inline test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/awsquerycompatible/AwsQueryCompatibleErrorTest.java:89]: Missing `fail()` in all test assertions, actually can we use `assertThatThrowBy` instead? 

Example:

https://github.com/aws/aws-sdk-java-v2/blob/5af85b1094334cdabd254fac175b0a789f01d556/services/s3/src/test/java/software/amazon/awssdk/services/s3/functionaltests/CompleteMultipartUploadFunctionalTest.java#L182-L185

Comment (ziyanli-amazon) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/awsquerycompatible/AwsQueryCompatibleErrorTest.java]: Cool. Refactored accordingly.

**Review by zoewangg — APPROVED** [review]
